web/profiles/custom/os2loop/modules/os2loop_flag_content/src/Helper/MailHelper.php

```diff
@@ -3,30 +3,112 @@
 
 use Drupal\Core\StringTranslation\StringTranslationTrait;
+use Drupal\Core\Utility\Token;
+use Drupal\os2loop_flag_content\Form\SettingsForm;
+use Drupal\os2loop_settings\Settings;
 
 /**
  * MailHelper for creating mail templates.
  */
 class MailHelper {
   use StringTranslationTrait;
+  /**
+   * The toke.
+   *
+   * @var \Drupal\Core\Utility\Token
+   */
+  protected $token;
+
+  /**
+   * The config.
+   *
+   * @var \Drupal\Core\Config\ImmutableConfig
+   */
+  private $config;
+
+  /**
+   * Constructor.
+   */
+  public function __construct(Token $token, Settings $settings) {
+    $this->token = $token;
+    $this->config = $settings->getConfig(SettingsForm::SETTINGS_NAME);
+  }
 
   /**
    * Implements hook_mail().
    */
   public function mail($key, &$message, $params) {
-    $options = [
-      'langcode' => $message['langcode'],
-    ];
     switch ($key) {
       case 'flag_content':
-        $subject = $this->t('Flagged content: @reason with @title', [
-          '@title' => $params['node_title'],
-          '@reason' => $params['reason'],
-        ], $options);
+        $node = $params['node'];
+        $body_template = $this->config->get('template_body');
+        $subject_template = $this->config->get('template_subject');
+        $data['node'] = $node;
+        $data['reason'] = $params['reason'];
+        $data['message'] = $params['message'];
+        $body = $this->renderTemplate($body_template, $data);
+        $subject = $this->renderTemplate($subject_template, $data);
         $message['subject'] = $subject;
-        $message['body'][] = $params['message'];
+        $message['body'][] = $body;
         break;
     }
   }
 
+  /**
+   * Renders content of a mail.
+   */
+  public function renderTemplate($template, array $data = NULL) {
+    return $this->token->replace($template, [
+      'node' => $data['node'],
+      'reason' => $data['reason'],
+      'message' => $data['message'],
+    ], []);
+
+  }
+
+  /**
+   * Implements hook_tokens().
+   */
+  public function tokens($type, $tokens, array $data) {
+    $replacements = [];
+    if ($type == 'os2loop_flag_content' && !empty($data['reason']) && !empty($data['message'])) {
+      foreach ($tokens as $name => $original) {
+        switch ($name) {
+          case 'reason':
+            $replacements[$original] = $data['reason'];
+            break;
+
+          case 'message':
+            $replacements[$original] = $data['message'];
+            break;
+        }
+      }
+    }
+    return $replacements;
+  }
+
+  /**
+   * Implements hook_token_info().
+   */
+  public function tokenInfo() {
+    $types['os2loop_flag_content'] = [
+      'name' => $this->t('Reason type'),
+    ];
+    $tokens['reason'] = [
+      'name' => $this->t('Reason'),
+    ];
+
+    $tokens['message'] = [
+      'name' => $this->t('Message'),
+    ];
+
+    return [
+      'types' => $types,
+      'tokens' => [
+        'os2loop_flag_content' => $tokens,
+      ],
+    ];
+
+  }
```

Comment:
This could be made a little simpler by adding our custom data under the `os2loop_flag_content` key in `$data`. See https://github.com/itk-dev/os2loop/blob/develop/web/profiles/custom/os2loop/modules/os2loop_share_with_a_friend/src/Helper/MailHelper.php#L64-L101 for an similar implementation.